File: pascal/2015/day08a/day08a.pas
PROGRAM Day08a(Input, Output);

VAR
    FileName : PChar;
    Res      : Integer;

PROCEDURE Usage();
VAR
    ProgName : PChar;
BEGIN
    ProgName := argv[0];
    WriteLn('usage: ', ProgName, ' <input file>');
    Halt(1)
END;

FUNCTION Process(FileName : PChar) : Integer;
VAR
    InFile          : Text;
    Res             : Integer;
    LineStr         : String;
    CodeLen, MemLen : Integer;
    I               : Integer;

BEGIN
    Res := 0;
    {$I-}
    Assign(InFile, FileName);
    Reset(InFile);
    IF IOResult <> 0 THEN
    BEGIN
        WriteLn(StdErr, 'error when opening file: ', FileName);
        Halt(1)
    END;
    {$I+}
    WHILE NOT EOF(InFile) DO
    BEGIN
        ReadLn(InFile, LineStr);
        CodeLen := Length(LineStr);
        MemLen := 0;
        I := 2;
        WHILE ( I < Length(LineStr) ) DO
        BEGIN
            IF LineStr[I] = '\' THEN
                CASE LineStr[I + 1] OF
                    '\', '"' : I := I + 2;
                    'x'      : I := I + 4;
                ELSE
                    I := I + 1;
                END
            ELSE
                I := I + 1;
            MemLen := MemLen + 1
        END;
        Res := Res + (CodeLen - MemLen)
    END;
    Close(InFile);
    Process := Res
END;

BEGIN
    IF argc < 2 THEN
        Usage();
    FileName := argv[1];
    Res := Process(FileName);
    WriteLn('result = ', Res)
END.

File: c++/2015/day08a/day08a.cpp
#include <cstdint>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <string>

void usage(std::string progname) {
  std::cerr << "usage: " << progname << " <input file>" << std::endl;
  std::exit(1);
}

uint32_t process(std::string filename) {
  uint32_t result = 0;
  std::ifstream infile(filename);
  std::string line;
  while (std::getline(infile, line)) {
    uint32_t code_len = line.length();
    uint32_t mem_len = 0;
    for (auto i = 1; i < line.length() - 1;) {
      switch (line[i]) {
        case '\\':
          switch (line[i + 1]) {
            case '\\':
            case '\"':
              i += 2;
              break;
            case 'x':
              i += 4;
              break;
            default:
              i++;
          }
          break;
        default:
          i++;
      }
      mem_len++;
    }
    result += code_len - mem_len;
  }
  return result;
}

int main(int argc, char *argv[]) {
  std::string progname(argv[0]);
  if (argc < 2) {
    usage(progname);
  }
  std::string filename(argv[1]);
  uint32_t result = process(filename);
  std::cout << "result = " << result << std::endl;
  return 0;
}
